### PR TITLE
chore: migration from SurfaceTextureEntry to SurfaceProducer

### PR DESCRIPTION
**Pull request by abdelaziz-mahdy**
this issue is a fix for https://github.com/media-kit/media-kit/issues/707 , since flutter migrated to SurfaceProducer for impeller https://docs.flutter.dev/release/breaking-changes/android-surface-plugins 

right now this is blocked by setOnFrameAvailableListener , since i cant find it in the new class and cant find any mention of it anywhere

### CODE REVIEW

**Comment by abdelaziz-mahdy** [comment]
i did try https://developer.android.com/reference/android/view/SurfaceHolder either that i am doing it wrong due to not understanding the code fully, or it doesnt work for our case anyway, leaving this here for tracking https://github.com/abdelaziz-mahdy/media_kit/tree/migration-to-impller-test

**Comment by abdelaziz-mahdy** [comment]
also i tested with `Choreographer` in https://github.com/abdelaziz-mahdy/media_kit/tree/migration-to-impller-Choreographer


both branches have the same problem, the first video rendering will be incorrect, and then the next video will be correct,note the `video 1 and 2 works` maybe the switch gives it time, i dont know , and there is a black screen that happens when switching alot of times  and its error is 

```
I/media_kit(12018): com.alexmercerind.media_kit_video.VideoOutputManager.createSurface: -5476376647959460208
W/System.err(12018): java.lang.IllegalStateException: Surface has already been released.
W/System.err(12018): 	at android.view.Surface.checkNotReleasedLocked(Surface.java:757)
W/System.err(12018): 	at android.view.Surface.lockCanvas(Surface.java:435)
W/System.err(12018): 	at com.alexmercerind.media_kit_video.VideoOutput.clearSurface(VideoOutput.java:195)
W/System.err(12018): 	at com.alexmercerind.media_kit_video.VideoOutput.createSurface(VideoOutput.java:163)
W/System.err(12018): 	at com.alexmercerind.media_kit_video.VideoOutputManager.createSurface(VideoOutputManager.java:55)
W/System.err(12018): 	at com.alexmercerind.media_kit_video.MediaKitVideoPlugin.onMethodCall(MediaKitVideoPlugin.java:115)
W/System.err(12018): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:267)
W/System.err(12018): 	at io.flutter.embedding.engine.dart.DartMessenger.invokeHandler(DartMessenger.java:292)
W/System.err(12018): 	at io.flutter.embedding.engine.dart.DartMessenger.lambda$dispatchMessageToQueue$0$io-flutter-embedding-engine-dart-DartMessenger(DartMessenger.java:319)
W/System.err(12018): 	at io.flutter.embedding.engine.dart.DartMessenger$$ExternalSyntheticLambda0.run(Unknown Source:12)
W/System.err(12018): 	at android.os.Handler.handleCallback(Handler.java:938)
W/System.err(12018): 	at android.os.Handler.dispatchMessage(Handler.java:99)
W/System.err(12018): 	at android.os.Looper.loopOnce(Looper.java:201)
W/System.err(12018): 	at android.os.Looper.loop(Looper.java:288)
W/System.err(12018): 	at android.app.ActivityThread.main(ActivityThread.java:7839)
W/System.err(12018): 	at java.lang.reflect.Method.invoke(Native Method)
W/System.err(12018): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
W/System.err(12018): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```

example of rendering problems 
the video will flicker only colors 
image

the video scale is incorrect
image

yes i am testing on emulator but on main none of these problems happen so there is a problem in the migration i did, @alexmercerind looks like i failed to do and will need your help in this one

**Comment by abdelaziz-mahdy** [comment]
cant figure out how to fix those

the first video render as a flicker of colors

black screen instead of rendering, it doesnt happen all the time, so i cant figure when and why it happens

video scale is wrong(this is gone, or much less) so lets ignore it for now (only for the first video from the sample it happens)

**Comment by abdelaziz-mahdy** [comment]
still faces the same problems above :_(

**Comment by abdelaziz-mahdy** [comment]
after alot of help from vith on discord its much more stable

File: media_kit_test/android/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -30,6 +30,9 @@
         <meta-data
             android:name="flutterEmbedding"
             android:value="2" />
+            <meta-data
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply media_kit_test/android/app/src/main/AndroidManifest.xml:33]: Yes it works with impeller

File: media_kit_video/lib/src/video_controller/android_video_controller/real.dart [hunk 2]
```diff
@@ -61,8 +62,8 @@ class AndroidVideoController extends PlatformVideoController {
       debugPrint('media_kit: AndroidVideoController: Enforcing S/W rendering.');
       enableHardwareAcceleration = false;
     }
-    _hwdec = configuration.hwdec ??
-        (enableHardwareAcceleration ? 'auto-safe' : 'no');
+    _hwdec =
+        configuration.hwdec ?? (enableHardwareAcceleration ? 'auto' : 'no');
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply media_kit_video/lib/src/video_controller/android_video_controller/real.dart:66]: Guess more stability is better? What do you think?

File: media_kit_video/lib/src/video_controller/android_video_controller/real.dart [hunk 3]
```diff
@@ -78,80 +79,12 @@ class AndroidVideoController extends PlatformVideoController {
...
-          final current = path.cast<Utf8>().toDartString();
-          calloc.free(name.cast());
-          mpv.mpv_free(path.cast());
-
-          bool refresh = _current != current;
-
-          _current = current;
-
-          if (refresh) {
-            // It is important to use a new android.view.Surface each time a new video-output is created because: https://stackoverflow.com/a/21564236
```
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply media_kit_video/lib/src/video_controller/android_video_controller/real.dart:103]: Set callback should update the surface I guess no refresh is needed

File: README.md
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply README.md:1]: <3

File: media_kit_video/lib/src/video_controller/android_video_controller/real.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (abdelaziz-mahdy) [reply media_kit_video/lib/src/video_controller/android_video_controller/real.dart:103]: can you elaborate when it happens so i can test the problem if it happens or not?

**Comment by abdelaziz-mahdy** [comment]
due to errors on flutter repo i decided to update the min flutter sdk 

https://github.com/flutter/flutter/issues/154054
https://github.com/flutter/flutter/issues/154050

i think thats the safest route

**Comment by abdelaziz-mahdy** [comment]
@alexmercerind  @moffatman  what do you guys think is the best option?

**Comment by moffatman** [comment]
bumping sdk constraint is breaking, should be a major

**Comment by abdelaziz-mahdy** [comment]
> bumping sdk constraint is breaking, should be a major

updated to version 2.0.0

**Comment by abdelaziz-mahdy** [comment]
@alexmercerind i saw you self requested a review on this,  do you think its needs work before merging or we merge it and publish it

**Comment by alexmercerind** [comment]
Hi!

This is still on my radar. Please bear with me.

I also aim to get rid of media_kit_native_event_loop in this version. There have been many issues with it (primarily on Window, as we have seen... deadlocks, breaking ABI change in MSVCP etc.). On the other hand... it was a workaround at that time, which I always wanted to remove.

Since we are changing the SDK constraint... why not migrate to https://api.flutter.dev/flutter/dart-ffi/NativeCallable-class.html.

**Comment by abdelaziz-mahdy** [comment]
oh got you i think you can use this as a reference  https://github.com/rainyl/opencv_dart/blob/cbfcc9e614e8f37fa608f22cc498aec147bea056/lib/src/core/base.dart#L92 we did use NativeCallable there to make the async functions, i only helped with the migration so i dont understand it fully to migrate media_kit_native_event_loop stuff :_(

**Comment by alexmercerind** [comment]
Thanks a lot for this pull-request... No doubt, I have taken many references from it.

However, I decided to proceed with:

- #962

There were a lot of things which could be simplified & resolved few other things which were not working.

**Pull request closed without merging**